### PR TITLE
Extern fees estimator

### PR DESCRIPTION
**Pull request by nearmax**

### CODE REVIEW

**Comment by icerove** [comment]
do you think initial balance should be at least one near token, it will be 10^18, or at now only 10^15 is set to be?

File: genesis-tools/genesis-populate/src/lib.rs [hunk 1]
```diff
@@ -238,17 +238,17 @@ impl GenesisBuilder {
     }
 
     fn add_additional_account(&mut self, account_id: AccountId) -> Result<()> {
-        const TESTING_INIT_BALANCE: Balance = 1_000_000_000_000_000;
-        const TESTING_INIT_STAKE: Balance = 0;
+        let testing_init_balance: Balance = 10u128.pow(30);
```
Comment (ailisp) [inline genesis-tools/genesis-populate/src/lib.rs]: Looks like need to rerun the 80M account generation. it's 75M/80M accounts now :joy:

Comment (nearmax) [reply genesis-tools/genesis-populate/src/lib.rs]: Unless you will be re-using the same accounts for lots of very heavy computation you don't need to.

File: runtime/runtime-params-estimator/src/cases.rs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+use rand::seq::SliceRandom;
+use rand::Rng;
+use std::cell::RefCell;
+use std::collections::{HashMap, HashSet};
+use std::path::PathBuf;
+
+pub fn run(config: Config) {
+    let mut m = Measurements::new();
+    // Measure the speed of processing empty receipts.
+    measure_actions("receipt only", 1, None, &mut m, &config, None, vec![], false, false);
```
Comment (ailisp) [inline runtime/runtime-params-estimator/src/cases.rs]: Why most of actions is measured only once?

File: runtime/runtime-params-estimator/src/cases.rs [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+            account_id.clone(),
+            &signer,
+            vec![function_call],
+            CryptoHash::default(),
+        )
+    };
+    testbed = measure_transactions("call noop", 1, None, &mut m, &config, Some(testbed), &mut f);
+
+    // Measure the speed of all extern function calls.
+    testbed = measure_function("call_fixture10", 10, &mut m, testbed, &ad, &mut nonces, &config);
```
Comment (ailisp) [inline runtime/runtime-params-estimator/src/cases.rs]: ... and these functions are measured 10/20/100 times. I mean how to figure out how many times makes sense to measure

File: runtime/runtime-params-estimator/src/main.rs [hunk 4]
```diff
@@ -1,312 +1,74 @@
...
+
+    let state_dump_path = matches.value_of("home").unwrap().to_string();
+    let warmup_iters_per_block = matches.value_of("warmup-iters").unwrap().parse().unwrap();
+    let iter_per_block = matches.value_of("iters").unwrap().parse().unwrap();
+    let active_accounts = matches.value_of("accounts-num").unwrap().parse().unwrap();
+    let smallest_block_size_pow2 =
+        matches.value_of("smallest-block-size-pow2").unwrap().parse().unwrap();
+    let largest_block_size_pow2 =
+        matches.value_of("largest-block-size-pow2").unwrap().parse().unwrap();
+    run(Config {
```
Comment (ailisp) [inline runtime/runtime-params-estimator/src/main.rs]: Cool all parameter is configurable now

File: runtime/runtime-params-estimator/src/cases.rs (hunk 2 above)
Comment (nearmax) [reply runtime/runtime-params-estimator/src/cases.rs]: This number is how many times an operation is executed per transaction. It is tricky to execute many actions per single transaction, because we might be doing some optimization when there are multiple actions of the same kind inside a single transaction, also certain sequences are prohibited, e.g. we cannot create an account twice within a single transaction. Also, if create a transaction with multiple actions each of which is doing a function call the receipts produced by all function calls except the last one are going to be ingnored. So for simplicity we measure each action once per transaction, but then we have multiple transactions per block and we can figure out the overhead of how much an action costs by itself by subtracting the cost/time of processing an empty transaction.

File: runtime/runtime-params-estimator/test-contract/build.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+# First, measure the size of the file without payload.
+rm -rf target
+RUSTFLAGS='-C link-arg=-s' cargo +nightly build --target wasm32-unknown-unknown --release
+bare_wasm=$(stat -c%s target/wasm32-unknown-unknown/release/test_contract.wasm)
+echo ${bare_wasm}
+
+# Generate several files of various sizes. We will compile these files into the Wasm binary to
+# bloat its size to the given values.
```
Comment (ailisp) [inline runtime/runtime-params-estimator/test-contract/build.sh]: This is really cool :sunglasses:

File: runtime/runtime-params-estimator/test-contract/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,571 @@
+#![no_std]
```
Comment (ailisp) [inline runtime/runtime-params-estimator/test-contract/src/lib.rs]: i guess this file can be copy as part of build.sh, to avoid maintain server copy of it

File: runtime/runtime-params-estimator/src/cases.rs (hunk 3 above)
Comment (nearmax) [reply runtime/runtime-params-estimator/src/cases.rs]: So these magic numbers have the following meaning:
* The number tells the given contract function how many times to run a certain operation in a loop;
* We want this operation to be executed as many time as possible to get statistical precision, however certain operations are too slow/expensive to run 100 times. So I figured out the constants that are large enough but do not make the code really slow.

In the future, we should have a more scientific approach for these constants (CC @icerove ), but for now we need to at least figure out the order of magnitude for fees costs, so I think these constants are good enough for now to give us the ballpark.

File: runtime/runtime-params-estimator/test-contract/src/lib.rs (hunk 6 above)
Comment (nearmax) [reply runtime/runtime-params-estimator/test-contract/src/lib.rs]: I am not sure I understand what you mean. Could you rephrase it?

File: runtime/runtime-params-estimator/src/cases.rs (hunk 2 above)
Comment (ailisp) [reply runtime/runtime-params-estimator/src/cases.rs]: I see.

> also certain sequences are prohibited e.g. we cannot create an account twice within a single transaction

These basically cause only one op/txn make sense.

File: runtime/runtime-params-estimator/src/cases.rs (hunk 3 above)
Comment (ailisp) [reply runtime/runtime-params-estimator/src/cases.rs]: Is this the case of

> Also, if create a transaction with multiple actions each of which is doing a function call the receipts produced by all function calls except the last one are going to be ingnored.

**Comment by nearmax** [comment]
> do you think initial balance should be at least one near token, it will be 10^18, or at now only 10^15 is set to be?

For testing purposes, I would populate accounts with almost max possible amount of tokens, in this PR we set it to 10**30. In mainnet, I don't think we should have any rules on what balance account should have unless we are talking about the case when the account runs out of balance due to rent.

> it will be 10^18,

I am not sure I remember if this an agreed number. Did we discuss it somewhere? I tried making some crude estimates in the past here: https://docs.google.com/document/d/1rydPgSHwPaKQLL57Xogsw8tgjTPKG98cK3kgK_PnjCQ/edit#heading=h.a9tzccsro77v But I think after running economics parameter estimation we will have a much better grasp on what numbers should be like, at least when we talk about orders of magnitude.

File: runtime/runtime-params-estimator/src/cases.rs (hunk 3 above)
Comment (nearmax) [reply runtime/runtime-params-estimator/src/cases.rs]: No, I am not sure what similarity are you trying to point to. Here we create transactions where each transaction has one action and this action is a function call to the contract, but inside this contract there is a loop that repeats certain action several times. This is very different than having multiple actions (where each action is a function call) in one transactions.

**Comment by ailisp** [comment]
I see. this is n txns, each txn has only one action and this action is a
function call. What's the use case for one txn multiple action? is it for
atomic operation?

**Comment by nearmax** [comment]
> What's the use case for one txn multiple action? is it for
atomic operation?

We don't measure this yet in this tool. The general use case is to have a transaction with the following actions:
* create account;
* add access key;
* deploy smart contract to it;
* call a function that would do initialization of that smart contract.
